### ISSUE TITLE
pubsub: Unable to terminate process

### ISSUE THREAD

**Issue by christhompsongoogle** [firebase/firebase-tools#5252]
firebase emulators:start --debug with pubsub enabled, followed by Ctrl + C.

Pubsub regularly fails to shutdown and needs to killed manually before restarting the emulators.

output:

```
i  emulators: Received SIGINT (Ctrl-C) for the first time. Starting a clean shutdown. 
i  emulators: Please wait for a clean shutdown or send the SIGINT (Ctrl-C) signal again to stop right now. 
i  emulators: Shutting down emulators. {"metadata":{"emulator":{"name":"hub"},"message":"Shutting down emulators."}}
i  ui: Stopping Emulator UI {"metadata":{"emulator":{"name":"ui"},"message":"Stopping Emulator UI"}}
⚠  Emulator UI has exited upon receiving signal: SIGINT 
i  extensions: Stopping Extensions Emulator {"metadata":{"emulator":{"name":"extensions"},"message":"Stopping Extensions Emulator"}}
[2022-11-16T23:15:19.201Z] [Extensions] Stopping Extensions emulator, this is a noop.
i  functions: Stopping Functions Emulator {"metadata":{"emulator":{"name":"functions"},"message":"Stopping Functions Emulator"}}
i  hosting: Stopping Hosting Emulator {"metadata":{"emulator":{"name":"hosting"},"message":"Stopping Hosting Emulator"}}
i  database: Stopping Database Emulator {"metadata":{"emulator":{"name":"database"},"message":"Stopping Database Emulator"}}
[2022-11-16T23:15:19.207Z] 15:15:19.206 [Thread-0] INFO com.firebase.server.forge.App$ - Attempting graceful shutdown.
 {"metadata":{"emulator":{"name":"database"},"message":"15:15:19.206 [Thread-0] INFO com.firebase.server.forge.App$ - Attempting graceful shutdown.\n"}}
[2022-11-16T23:15:19.214Z] 15:15:19.214 [NamespaceSystem-akka.actor.default-dispatcher-4] INFO com.firebase.core.namespace.Terminator$Terminator - 1 actors left to terminate: isdhaksdflashgfal-default-rtdb
 {"metadata":{"emulator":{"name":"database"},"message":"15:15:19.214 [NamespaceSystem-akka.actor.default-dispatcher-4] INFO com.firebase.core.namespace.Terminator$Terminator - 1 actors left to terminate: isdhaksdflashgfal-default-rtdb\n"}}
[2022-11-16T23:15:19.225Z] 15:15:19.225 [NamespaceSystem-akka.actor.default-dispatcher-4] INFO com.firebase.core.namespace.NamespaceActor - stopped namespace actor for isdhaksdflashgfal-default-rtdb
 {"metadata":{"emulator":{"name":"database"},"message":"15:15:19.225 [NamespaceSystem-akka.actor.default-dispatcher-4] INFO com.firebase.core.namespace.NamespaceActor - stopped namespace actor for isdhaksdflashgfal-default-rtdb\n"}}
[2022-11-16T23:15:19.234Z] 15:15:19.233 [Thread-0] INFO com.firebase.server.forge.App$ - Graceful shutdown complete.
 {"metadata":{"emulator":{"name":"database"},"message":"15:15:19.233 [Thread-0] INFO com.firebase.server.forge.App$ - Graceful shutdown complete.\n"}}
i  firestore: Stopping Firestore Emulator {"metadata":{"emulator":{"name":"firestore"},"message":"Stopping Firestore Emulator"}}
[2022-11-16T23:15:19.266Z] *** shutting down gRPC server since JVM is shutting down {"metadata":{"emulator":{"name":"firestore"},"message":"*** shutting down gRPC server since JVM is shutting down"}}
[2022-11-16T23:15:19.266Z] 
 {"metadata":{"emulator":{"name":"firestore"},"message":"\n"}}
[2022-11-16T23:15:19.278Z] *** server shut down
 {"metadata":{"emulator":{"name":"firestore"},"message":"*** server shut down\n"}}
i  pubsub: Stopping Pub/Sub Emulator {"metadata":{"emulator":{"name":"pubsub"},"message":"Stopping Pub/Sub Emulator"}}
[2022-11-16T23:15:23.290Z] Pub/Sub Emulator: Unable to terminate process (PID=1918054) {"metadata":{"emulator":{"name":"pubsub"},"message":"Pub/Sub Emulator: Unable to terminate process (PID=1918054)"}}
⚠  pubsub: Error stopping Pub/Sub Emulator {"metadata":{"emulator":{"name":"pubsub"},"message":"Error stopping Pub/Sub Emulator"}}
i  auth: Stopping Authentication Emulator {"metadata":{"emulator":{"name":"auth"},"message":"Stopping Authentication Emulator"}}
i  storage: Stopping Storage Emulator {"metadata":{"emulator":{"name":"storage"},"message":"Stopping Storage Emulator"}}
i  eventarc: Stopping Eventarc Emulator {"metadata":{"emulator":{"name":"eventarc"},"message":"Stopping Eventarc Emulator"}}
i  hub: Stopping emulator hub {"metadata":{"emulator":{"name":"hub"},"message":"Stopping emulator hub"}}
i  logging: Stopping Logging Emulator {"metadata":{"emulator":{"name":"logging"},"message":"Stopping Logging Emulator"}}
```

**Comment by jonathanquaade** [comment]
Got the same problem

**Comment by sue445** [comment]
I'm seeing same problem with ` firebase emulators:exec --only pubsub` too.

**Comment by JCMais** [comment]
I have the same issue, I ended up creating a script to trap sigint and then kill the process directly:

```sh
#!/usr/bin/env bash
PID="$$"

echo "The pubsub script PID is $PID"

trap "npx kill-port 8085 && kill -s KILL $PID" INT

exec gcloud beta emulators pubsub start --project=my-pubsub-project --host-port=localhost:8085
```

**Comment by stevgouws** [comment]
Same issue, this seems to have been introduced in`11.16.1` which added pubsub emulator `v0.7.1`

**Comment by just7mile** [comment]
Same issue for me, and also, it fails every single time - it is really annoying.
Could it be something with SLF4J? This is what is inside the `pubsub-debug.log`:

```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

_Note:_ Firebase version is 11.19.0.

**Comment by NeverwinterMoon** [comment]
Experiencing exactly the same issue after finally updating from 11.4.2 to 11.19.0. Was not an issue before that.

**Comment by christhompsongoogle** [comment]
> Could it be something with SLF4J? This is what is inside the pubsub-debug.log

Possibly; one interesting observation is that it doesn't seem to cause a problem when launched/killed from the gcloud side.

**Comment by aaga** [comment]
I'm seeing the same issue with all 3 of auth/functions/firestore enabled (i.e. not pubsub)

**Comment by christhompsongoogle** [comment]
Fixed in https://github.com/firebase/firebase-tools/pull/5294

**Comment by just7mile** [comment]
Thank you! But the fix has one issue though - when there are multiple Firebase Emulator Suite instances, exiting one of them will kill the PubSub emulators for all of them.

> Fixed in #5294

**Comment by bgazzera** [comment]
For anyone coming at this issue when using an Alpine-based linux distro (most Docker images), the #5294 fix won't work because of the way `ps aux` works in Alpine (it reverses the pid & user columns in the output).

When terminating the pubsub emulator you will see a message like:
```
Stopping Pub/Sub Emulator 
sh: invalid number 'root'
sh: invalid number 'root'
```

A solution is to install the `procps` package, like so: `apk --no-cache add procps`

**Comment by just7mile** [comment]
> Thank you! But the fix has one issue though - when there are multiple Firebase Emulator Suite instances, exiting one of them will kill the PubSub emulators for all of them.
> 
> > Fixed in #5294

Also, this causes other Emulator Suites to fail and do not export data (so changes are lost).
